File: backend-node/src/models/StreamingError.ts
import { ProviderType } from './ProviderType';

export abstract class StreamingError extends Error {
  constructor(message: string, public cause?: unknown) {
    super(message);
    this.name = 'StreamingError';
  }
}

export class ProviderError extends StreamingError {
  constructor(public provider: ProviderType, errorMessage: string, cause?: unknown) {
    super(`Provider error from ${provider}: ${errorMessage}`, cause);
    this.name = 'ProviderError';
  }
}

export class AuthenticationError extends StreamingError {
  constructor(public provider: ProviderType, cause?: unknown) {
    super(`Authentication failed for provider ${provider}`, cause);
    this.name = 'AuthenticationError';
  }
}

export class NetworkError extends StreamingError {
  constructor(message: string, cause?: unknown) {
    super(`Network error: ${message}`, cause);
    this.name = 'NetworkError';
  }
}

export class TrackNotFoundError extends StreamingError {
  constructor(public trackId: string, public provider?: ProviderType) {
    super(`Track not found: ${trackId}${provider ? ` on ${provider}` : ''}`);
    this.name = 'TrackNotFoundError';
  }
}

export class InvalidSeekPosition extends StreamingError {
  constructor(public position: number, public trackDuration: number) {
    super(`Invalid seek position ${position} (track duration ${trackDuration})`);
    this.name = 'InvalidSeekPosition';
  }
}

export class QueueError extends StreamingError {
  constructor(message: string) {
    super(`Queue error: ${message}`);
    this.name = 'QueueError';
  }
}

export class RateLimitError extends StreamingError {
  constructor(public provider: ProviderType, public retryAfterSeconds: number) {
    super(`Rate limit exceeded for ${provider}, retry after ${retryAfterSeconds}s`);
    this.name = 'RateLimitError';
  }
}

File: backend/src/main/kotlin/com/musicstreaming/models/StreamingError.kt
package com.musicstreaming.models

sealed class StreamingError(
    override val message: String,
    override val cause: Throwable? = null
) : Exception(message, cause) {
    
    data class ProviderError(
        val provider: ProviderType,
        val errorMessage: String,
        val errorCause: Throwable? = null
    ) : StreamingError("Provider error from ${provider.name}: $errorMessage", errorCause)
    
    data class AuthenticationError(
        val provider: ProviderType,
        val errorCause: Throwable? = null
    ) : StreamingError("Authentication failed for provider: ${provider.name}", errorCause)
    
    data class NetworkError(
        val errorMessage: String,
        val errorCause: Throwable? = null
    ) : StreamingError("Network error: $errorMessage", errorCause)
    
    data class TrackNotFoundError(
        val trackId: String,
        val provider: ProviderType? = null
    ) : StreamingError("Track not found: $trackId${provider?.let { " on $it" } ?: ""}")
    
    data class InvalidSeekPosition(
        val position: Long,
        val trackDuration: Long
    ) : StreamingError("Invalid seek position: $position (track duration: $trackDuration)")
    
    data class QueueError(
        val errorMessage: String
    ) : StreamingError("Queue error: $errorMessage")
    
    data class RateLimitError(
        val provider: ProviderType,
        val retryAfterSeconds: Long
    ) : StreamingError("Rate limit exceeded for ${provider.name}, retry after $retryAfterSeconds seconds")
}
